File: lib/domain/analytics.ts
import type { HoleScore, RoundRecord, RoundSummary, RoundType } from "@/lib/domain/types";

function average(values: number[]) {
  if (values.length === 0) return null;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function averageForDefined(values: Array<number | null>) {
  return average(values.filter((value): value is number => value != null));
}

export type AnalyticsSnapshot = {
  roundCount: number;
  bestScore: number | null;
  averageScore: number | null;
  recentFiveAverage: number | null;
  averagePutts: number | null;
  puttSampleCount: number;
  girRate: number | null;
  girSampleCount: number;
  fwRate: number | null;
  fwSampleCount: number;
  penaltyAverage: number | null;
  penaltySampleCount: number;
};

function completedRounds(records: RoundRecord[], roundType: RoundType) {
  return records.filter(
    ({ round }) => round.status === "COMPLETED" && round.roundType === roundType,
  );
}

export function calculateAnalytics(
  records: RoundRecord[],
  roundType: RoundType,
): AnalyticsSnapshot {
  const targetRounds = completedRounds(records, roundType).sort((a, b) =>
    b.round.playedAt.localeCompare(a.round.playedAt),
  );

  const totalScores = targetRounds
    .map(({ round }) => round.totalScore)
    .filter((score): score is number => score != null);

  const allHoles = targetRounds.flatMap(({ holes }) => holes);
  const girHoles = allHoles.filter((hole) => hole.gir != null);
  const fwHoles = allHoles.filter(
    (hole) => hole.fwKeep != null && hole.par >= 4,
  );

  return {
    roundCount: targetRounds.length,
    bestScore: totalScores.length ? Math.min(...totalScores) : null,
    averageScore: average(totalScores),
    recentFiveAverage: average(totalScores.slice(0, 5)),
    averagePutts: averageForDefined(allHoles.map((hole) => hole.putts)),
    puttSampleCount: allHoles.filter((hole) => hole.putts != null).length,
    girRate: girHoles.length
      ? girHoles.filter((hole) => hole.gir === true).length / girHoles.length
      : null,
    girSampleCount: girHoles.length,
    fwRate: fwHoles.length
      ? fwHoles.filter((hole) => hole.fwKeep === true).length / fwHoles.length
      : null,
    fwSampleCount: fwHoles.length,
    penaltyAverage: average(allHoles.map((hole) => hole.penaltyTotal)),
    penaltySampleCount: allHoles.length,
  };
}

export function toRoundSummaries(records: RoundRecord[]): RoundSummary[] {
  return records
    .map(({ round }) => ({
      id: round.id,
      playedAt: round.playedAt,
      courseId: round.courseId ?? null,
      courseName: round.courseName,
      roundType: round.roundType,
      status: round.status,
      totalScore: round.totalScore,
      lastInputAt: round.lastInputAt,
    }))
    .sort((a, b) => b.playedAt.localeCompare(a.playedAt));
}

export function calcRoundTotal(holes: HoleScore[]) {
  const filled = holes
    .map((hole) => hole.strokes)
    .filter((value): value is number => value != null);

  if (filled.length !== holes.length) return null;
  return filled.reduce((sum, value) => sum + value, 0);
}

File: flutter_app/lib/domain/analytics.dart
import 'models.dart';

double? _average(List<num> values) {
  if (values.isEmpty) return null;
  final total = values.fold<num>(0, (sum, value) => sum + value);
  return total / values.length;
}

double? _averageForDefined(List<int?> values) {
  final filtered = values.whereType<int>().toList();
  return _average(filtered);
}

bool? _isGir(HoleScore hole) {
  final strokes = hole.strokes;
  final putts = hole.putts;
  if (strokes == null || putts == null) return null;
  return strokes - putts <= hole.par - 2;
}

class AnalyticsSnapshot {
  const AnalyticsSnapshot({
    required this.roundCount,
    required this.bestScore,
    required this.averageScore,
    required this.recentFiveAverage,
    required this.averagePutts,
    required this.puttSampleCount,
    required this.girRate,
    required this.girSampleCount,
    required this.fwRate,
    required this.fwSampleCount,
    required this.penaltyAverage,
    required this.penaltySampleCount,
  });

  final int roundCount;
  final int? bestScore;
  final double? averageScore;
  final double? recentFiveAverage;
  final double? averagePutts;
  final int puttSampleCount;
  final double? girRate;
  final int girSampleCount;
  final double? fwRate;
  final int fwSampleCount;
  final double? penaltyAverage;
  final int penaltySampleCount;
}

List<RoundRecord> _completedRounds(
    List<RoundRecord> records, RoundType roundType) {
  return records.where((record) {
    return record.round.status == RoundStatus.completed &&
        record.round.roundType == roundType;
  }).toList();
}

AnalyticsSnapshot calculateAnalytics(
    List<RoundRecord> records, RoundType roundType) {
  final targetRounds = _completedRounds(records, roundType)
    ..sort((a, b) => b.round.playedAt.compareTo(a.round.playedAt));

  final totalScores = targetRounds
      .map((record) => record.round.totalScore)
      .whereType<int>()
      .toList();

  final allHoles = targetRounds.expand((record) => record.holes).toList();
  final girValues = allHoles.map(_isGir).whereType<bool>().toList();
  final fwHoles =
      allHoles.where((hole) => hole.fwKeep != null && hole.par >= 4).toList();

  return AnalyticsSnapshot(
    roundCount: targetRounds.length,
    bestScore: totalScores.isEmpty
        ? null
        : totalScores.reduce((a, b) => a < b ? a : b),
    averageScore: _average(totalScores),
    recentFiveAverage: _average(totalScores.take(5).toList()),
    averagePutts:
        _averageForDefined(allHoles.map((hole) => hole.putts).toList()),
    puttSampleCount: allHoles.where((hole) => hole.putts != null).length,
    girRate: girValues.isEmpty
        ? null
        : girValues.where((isGir) => isGir).length / girValues.length,
    girSampleCount: girValues.length,
    fwRate: fwHoles.isEmpty
        ? null
        : fwHoles.where((hole) => hole.fwKeep == true).length / fwHoles.length,
    fwSampleCount: fwHoles.length,
    penaltyAverage:
        _average(allHoles.map((hole) => hole.penaltyTotal).toList()),
    penaltySampleCount: allHoles.length,
  );
}

List<RoundSummary> toRoundSummaries(List<RoundRecord> records) {
  final summaries = records
      .map(
        (record) => RoundSummary(
          id: record.round.id,
          playedAt: record.round.playedAt,
          courseId: record.round.courseId,
          courseName: record.round.courseName,
          roundType: record.round.roundType,
          status: record.round.status,
          totalScore: record.round.totalScore,
          lastInputAt: record.round.lastInputAt,
        ),
      )
      .toList();

  summaries.sort((a, b) => b.playedAt.compareTo(a.playedAt));
  return summaries;
}

int? calcRoundTotal(List<HoleScore> holes) {
  final filled = holes.map((hole) => hole.strokes).whereType<int>().toList();
  if (filled.length != holes.length) return null;
  return filled.fold<int>(0, (sum, value) => sum + value);
}
